# Language Translation
In this project, you’re going to take a peek into the realm of neural network machine translation.  You’ll be training a sequence to sequence model on a dataset of English and French sentences that can translate new sentences from English to French.
## Get the Data
Since translating the whole language of English to French will take lots of time to train, we have provided you with a small portion of the English corpus.

In [1]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import helper
import problem_unittests as tests

source_path = 'data/small_vocab_en'
target_path = 'data/small_vocab_fr'
source_text = helper.load_data(source_path)
target_text = helper.load_data(target_path)

## Explore the Data
Play around with view_sentence_range to view different parts of the data.

In [2]:
view_sentence_range = (0, 10)

"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np

print('Dataset Stats')
print('Roughly the number of unique words: {}'.format(len({word: None for word in source_text.split()})))

sentences = source_text.split('\n')
word_counts = [len(sentence.split()) for sentence in sentences]
print('Number of sentences: {}'.format(len(sentences)))
print('Average number of words in a sentence: {}'.format(np.average(word_counts)))

print()
print('English sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(source_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))
print()
print('French sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(target_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))

Dataset Stats
Roughly the number of unique words: 227
Number of sentences: 137861
Average number of words in a sentence: 13.225277634719028

English sentences 0 to 10:
new jersey is sometimes quiet during autumn , and it is snowy in april .
the united states is usually chilly during july , and it is usually freezing in november .
california is usually quiet during march , and it is usually hot in june .
the united states is sometimes mild during june , and it is cold in september .
your least liked fruit is the grape , but my least liked is the apple .
his favorite fruit is the orange , but my favorite is the grape .
paris is relaxing during december , but it is usually chilly in july .
new jersey is busy during spring , and it is never hot in march .
our least liked fruit is the lemon , but my least liked is the grape .
the united states is sometimes busy during january , and it is sometimes warm in november .

French sentences 0 to 10:
new jersey est parfois calme pendant l' automne 

## Implement Preprocessing Function
### Text to Word Ids
As you did with other RNNs, you must turn the text into a number so the computer can understand it. In the function `text_to_ids()`, you'll turn `source_text` and `target_text` from words to ids.  However, you need to add the `<EOS>` word id at the end of `target_text`.  This will help the neural network predict when the sentence should end.

You can get the `<EOS>` word id by doing:
```python
target_vocab_to_int['<EOS>']
```
You can get other word ids using `source_vocab_to_int` and `target_vocab_to_int`.

In [76]:
def text_to_ids(source_text, target_text, source_vocab_to_int, target_vocab_to_int):
    """
    Convert source and target text to proper word ids
    :param source_text: String that contains all the source text.
    :param target_text: String that contains all the target text.
    :param source_vocab_to_int: Dictionary to go from the source words to an id
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: A tuple of lists (source_id_text, target_id_text)
    """
    
    # Using the "problem_unittests.py" and "helper.py" functions as a guide...    
    source_text_lc = source_text.lower()
    target_text_lc = target_text.lower()

    #source_vocab_to_int, source_int_to_vocab = helper.create_lookup_tables(test_source_text)
    #target_vocab_to_int, target_int_to_vocab = helper.create_lookup_tables(test_target_text)
    
    source_lines = source_text_lc.split('\n')
    target_lines = target_text_lc.split('\n')
    
    source_ids = []
    target_ids = []
    
    # Split each line of the source into tokens (words)
    for line in source_lines:
        
        # For each non-empty source string
        if line:
            
            # Replace double spaces with single spaces
            line = line.replace('  ', ' ')
            
            # Split each line into a list of tokens (words)
            source_words = line.strip().split(' ')
            
            # Lookup the (int) index for each source word
            source_ints = [source_vocab_to_int[x] for x in source_words]
            
            source_ids.append(source_ints)

    # Do the same for target words
    for line in target_lines:
        if line:
            line = line.replace('  ', ' ')
            target_words = line.strip().split(' ')
            target_ints = [target_vocab_to_int[x] for x in target_words]
            
            # Add the required <EOS> token to the end of each target line
            target_ints.append(target_vocab_to_int['<EOS>'])
            
            target_ids.append(target_ints)
            
    return (source_ids, target_ids)

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_text_to_ids(text_to_ids)

Tests Passed


### Preprocess all the data and save it
Running the code cell below will preprocess all the data and save it to file.

In [77]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
helper.preprocess_and_save_data(source_path, target_path, text_to_ids)

# Check Point
This is your first checkpoint. If you ever decide to come back to this notebook or have to restart the notebook, you can start from here. The preprocessed data has been saved to disk.

In [78]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np
import helper
import problem_unittests as tests

(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()

### Check the Version of TensorFlow and Access to GPU
This will check to make sure you have the correct version of TensorFlow and access to a GPU

In [79]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
from distutils.version import LooseVersion
import warnings
import tensorflow as tf
from tensorflow.python.layers.core import Dense

# Check TensorFlow Version
assert LooseVersion(tf.__version__) >= LooseVersion('1.1'), 'Please use TensorFlow version 1.1 or newer'
print('TensorFlow Version: {}'.format(tf.__version__))

# Check for a GPU
if not tf.test.gpu_device_name():
    warnings.warn('No GPU found. Please use a GPU to train your neural network.')
else:
    print('Default GPU Device: {}'.format(tf.test.gpu_device_name()))

TensorFlow Version: 1.2.1


D:\Anaconda3\envs\TensorFlow\lib\site-packages\ipykernel_launcher.py:15: UserWarning: No GPU found. Please use a GPU to train your neural network.
  from ipykernel import kernelapp as app


## Build the Neural Network
You'll build the components necessary to build a Sequence-to-Sequence model by implementing the following functions below:
- `model_inputs`
- `process_decoder_input`
- `encoding_layer`
- `decoding_layer_train`
- `decoding_layer_infer`
- `decoding_layer`
- `seq2seq_model`

### Input
Implement the `model_inputs()` function to create TF Placeholders for the Neural Network. It should create the following placeholders:

- Input text placeholder named "input" using the TF Placeholder name parameter with rank 2.
- Targets placeholder with rank 2.
- Learning rate placeholder with rank 0.
- Keep probability placeholder named "keep_prob" using the TF Placeholder name parameter with rank 0.
- Target sequence length placeholder named "target_sequence_length" with rank 1
- Max target sequence length tensor named "max_target_len" getting its value from applying tf.reduce_max on the target_sequence_length placeholder. Rank 0.
- Source sequence length placeholder named "source_sequence_length" with rank 1

Return the placeholders in the following the tuple (input, targets, learning rate, keep probability, target sequence length, max target sequence length, source sequence length)

In [105]:
def model_inputs():
    """
    Create TF Placeholders for input, targets, learning rate, and lengths of source and target sequences.
    :return: Tuple (input, targets, learning rate, keep probability, target sequence length,
    max target sequence length, source sequence length)
    """
    
    # Input and target parameters
    inputs                 = tf.placeholder(tf.int32,   shape=(None, None), name='input'                 )
    targets                = tf.placeholder(tf.int32,   shape=(None, None), name='targets'               )
    
    # Training paramaeters
    learning_rate          = tf.placeholder(tf.float32, shape=(None),         name='learning_rate'         )
    keep_prob              = tf.placeholder(tf.float32, shape=(None),         name='keep_prob'             )
    
    # Sequence length parameters (source and target sequence lengths)
    target_sequence_length = tf.placeholder(tf.int32,   shape=(None,),       name='target_sequence_length')    
    source_sequence_length = tf.placeholder(tf.int32,   shape=(None,),       name='source_sequence_length')
    
    max_target_len         = tf.reduce_max (target_sequence_length)    
    
    return inputs, targets, learning_rate, keep_prob, target_sequence_length, max_target_len, source_sequence_length


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_model_inputs(model_inputs)

ERROR:tensorflow:==================================
Object was never used (type <class 'tensorflow.python.framework.ops.Operation'>):
<tf.Operation 'assert_rank_2/Assert/Assert' type=Assert>
If you want to mark it as used call its "mark_used()" method.
It was originally created here:
['File "D:\\Anaconda3\\envs\\TensorFlow\\lib\\runpy.py", line 193, in _run_module_as_main\n    "__main__", mod_spec)', 'File "D:\\Anaconda3\\envs\\TensorFlow\\lib\\runpy.py", line 85, in _run_code\n    exec(code, run_globals)', 'File "D:\\Anaconda3\\envs\\TensorFlow\\lib\\site-packages\\ipykernel_launcher.py", line 16, in <module>\n    app.launch_new_instance()', 'File "D:\\Anaconda3\\envs\\TensorFlow\\lib\\site-packages\\traitlets\\config\\application.py", line 658, in launch_instance\n    app.start()', 'File "D:\\Anaconda3\\envs\\TensorFlow\\lib\\site-packages\\ipykernel\\kernelapp.py", line 477, in start\n    ioloop.IOLoop.instance().start()', 'File "D:\\Anaconda3\\envs\\TensorFlow\\lib\\site-packages\\

Tests Passed


### Process Decoder Input
Implement `process_decoder_input` by removing the last word id from each batch in `target_data` and concat the GO ID to the begining of each batch.

In [106]:
def process_decoder_input(target_data, target_vocab_to_int, batch_size):
    """
    Preprocess target data for encoding
    :param target_data: Target Placehoder
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param batch_size: Batch Size
    :return: Preprocessed target data
    """
    
    # Remove the last word id from the target_data tensor
    last_word_removed = tf.strided_slice(input_  = target_data, 
                                         begin   = [0, 0], 
                                         end     = [batch_size, -1], 
                                         strides = [1, 1])
    
    # Add the "<GO>" ID to the beginning of each batch
    go_added = tf.fill([batch_size, 1], target_vocab_to_int['<GO>'])
    
    result = tf.concat([go_added, last_word_removed], 1)
    
    return result

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_process_encoding_input(process_decoder_input)

Tests Passed


### Encoding
Implement `encoding_layer()` to create a Encoder RNN layer:
 * Embed the encoder input using [`tf.contrib.layers.embed_sequence`](https://www.tensorflow.org/api_docs/python/tf/contrib/layers/embed_sequence)
 * Construct a [stacked](https://github.com/tensorflow/tensorflow/blob/6947f65a374ebf29e74bb71e36fd82760056d82c/tensorflow/docs_src/tutorials/recurrent.md#stacking-multiple-lstms) [`tf.contrib.rnn.LSTMCell`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/LSTMCell) wrapped in a [`tf.contrib.rnn.DropoutWrapper`](https://www.tensorflow.org/api_docs/python/tf/contrib/rnn/DropoutWrapper)
 * Pass cell and embedded input to [`tf.nn.dynamic_rnn()`](https://www.tensorflow.org/api_docs/python/tf/nn/dynamic_rnn)

In [107]:
from imp import reload
reload(tests)

def lstm_cell(lstm_size):
  return tf.contrib.rnn.BasicLSTMCell(lstm_size)


def lstm_with_dropout(lstm_size, keep_prob):
    return tf.contrib.rnn.DropoutWrapper(lstm_cell(lstm_size), 
                                         output_keep_prob = keep_prob)

def encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob, 
                   source_sequence_length, source_vocab_size, 
                   encoding_embedding_size):
    """
    Create encoding layer
    :param rnn_inputs: Inputs for the RNN
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param keep_prob: Dropout keep probability
    :param source_sequence_length: a list of the lengths of each sequence in the batch
    :param source_vocab_size: vocabulary size of source data
    :param encoding_embedding_size: embedding size of source data
    :return: tuple (RNN output, RNN state)
    """
    
    # Embedding inputs    
    embedding_inputs = tf.contrib.layers.embed_sequence(ids        = rnn_inputs, 
                                                        vocab_size = source_vocab_size, 
                                                        embed_dim  = encoding_embedding_size)
        
    # Create a stack of LSTMs (ref: https://www.tensorflow.org/tutorials/recurrent)
    stacked_lstm = tf.contrib.rnn.MultiRNNCell( [lstm_with_dropout(rnn_size, keep_prob) for _ in range(num_layers)])
    
    rnn_output, rnn_state = tf.nn.dynamic_rnn(cell            = stacked_lstm, 
                                              inputs          = embedding_inputs,
                                              sequence_length = source_sequence_length,
                                              dtype           = tf.float32)
    
    return (rnn_output, rnn_state)

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_encoding_layer(encoding_layer)

Tests Passed


### Decoding - Training
Create a training decoding layer:
* Create a [`tf.contrib.seq2seq.TrainingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/TrainingHelper) 
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [128]:
def decoding_layer_train(encoder_state, dec_cell, dec_embed_input, 
                         target_sequence_length, max_summary_length, 
                         output_layer, keep_prob):
    """
    Create a decoding layer for training
    :param encoder_state: Encoder State
    :param dec_cell: Decoder RNN Cell
    :param dec_embed_input: Decoder embedded input
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_summary_length: The length of the longest sequence in the batch
    :param output_layer: Function to apply the output layer
    :param keep_prob: Dropout keep probability
    :return: BasicDecoderOutput containing training logits and sample_id
    """
    
    # ref:  https://www.tensorflow.org/api_guides/python/contrib.seq2seq#Dynamic_Decoding
    
    helper = tf.contrib.seq2seq.TrainingHelper(
        inputs          = dec_embed_input,
        sequence_length = target_sequence_length,
        time_major      = False)

    decoder = tf.contrib.seq2seq.BasicDecoder(
        cell          = dec_cell,
        helper        = helper,
        initial_state = encoder_state,
        output_layer  = output_layer)
    
    outputs, _, _ = tf.contrib.seq2seq.dynamic_decode(
       decoder            = decoder,
       impute_finished    = True,
       maximum_iterations = max_summary_length)
    
    return outputs

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_train(decoding_layer_train)

Tests Passed



### Decoding - Inference
Create inference decoder:
* Create a [`tf.contrib.seq2seq.GreedyEmbeddingHelper`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/GreedyEmbeddingHelper)
* Create a [`tf.contrib.seq2seq.BasicDecoder`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/BasicDecoder)
* Obtain the decoder outputs from [`tf.contrib.seq2seq.dynamic_decode`](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/dynamic_decode)

In [130]:
def decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id,
                         end_of_sequence_id, max_target_sequence_length,
                         vocab_size, output_layer, batch_size, keep_prob):
    """
    Create a decoding layer for inference
    :param encoder_state: Encoder state
    :param dec_cell: Decoder RNN Cell
    :param dec_embeddings: Decoder embeddings
    :param start_of_sequence_id: GO ID
    :param end_of_sequence_id: EOS Id
    :param max_target_sequence_length: Maximum length of target sequences
    :param vocab_size: Size of decoder/target vocabulary
    :param decoding_scope: TenorFlow Variable Scope for decoding
    :param output_layer: Function to apply the output layer
    :param batch_size: Batch size
    :param keep_prob: Dropout keep probability
    :return: BasicDecoderOutput containing inference logits and sample_id
    """
    
    GO_SYMBOL  = target_vocab_to_int['<GO>']
    END_SYMBOL = target_vocab_to_int['<EOS>']
    
    helper = tf.contrib.seq2seq.GreedyEmbeddingHelper(
          embedding    = dec_embeddings,
          start_tokens = tf.tile([GO_SYMBOL], [batch_size]),
          end_token    = END_SYMBOL)

    # Define the decoder
    decoder = tf.contrib.seq2seq.BasicDecoder(
        cell          = dec_cell,
        helper        = helper,
        initial_state = encoder_state,
        output_layer  = output_layer)
    
    # Decode, using the decoder
    outputs, _, _ = tf.contrib.seq2seq.dynamic_decode(
       decoder            = decoder,
       impute_finished    = True,
       maximum_iterations = max_target_sequence_length)

    return outputs

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_infer(decoding_layer_infer)

Tests Passed


### Build the Decoding Layer
Implement `decoding_layer()` to create a Decoder RNN layer.

* Embed the target sequences
* Construct the decoder LSTM cell (just like you constructed the encoder cell above)
* Create an output layer to map the outputs of the decoder to the elements of our vocabulary
* Use the your `decoding_layer_train(encoder_state, dec_cell, dec_embed_input, target_sequence_length, max_target_sequence_length, output_layer, keep_prob)` function to get the training logits.
* Use your `decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id, end_of_sequence_id, max_target_sequence_length, vocab_size, output_layer, batch_size, keep_prob)` function to get the inference logits.

Note: You'll need to use [tf.variable_scope](https://www.tensorflow.org/api_docs/python/tf/variable_scope) to share variables between training and inference.

In [133]:
def decoding_layer(dec_input, encoder_state,
                   target_sequence_length, max_target_sequence_length,
                   rnn_size,
                   num_layers, target_vocab_to_int, target_vocab_size,
                   batch_size, keep_prob, decoding_embedding_size):
    """
    Create decoding layer
    :param dec_input: Decoder input
    :param encoder_state: Encoder state
    :param target_sequence_length: The lengths of each sequence in the target batch
    :param max_target_sequence_length: Maximum length of target sequences
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param target_vocab_size: Size of target vocabulary
    :param batch_size: The size of the batch
    :param keep_prob: Dropout keep probability
    :param decoding_embedding_size: Decoding embedding size
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """
      
    # Create the decoder embedding
    dec_embeddings   = tf.Variable(tf.random_uniform([target_vocab_size, decoding_embedding_size]))
    dec_embed_inputs = tf.nn.embedding_lookup(dec_embeddings, dec_input)

    # Create the decoding RNN cell. 
    lstm     = lambda rnn_size: tf.contrib.rnn.LSTMCell(rnn_size, 
                                                        initializer=tf.random_uniform_initializer(-0.1, 0.1))
    
    dec_cell = tf.contrib.rnn.MultiRNNCell([lstm(rnn_size) for _ in range(num_layers)])

    output_layer = Dense(units              = target_vocab_size,
                         kernel_initializer = tf.truncated_normal_initializer(mean = 0.0, stddev=0.1))
    
    with tf.variable_scope('training'): 
        
        # Create decoding layer train
        train_pred = decoding_layer_train(encoder_state,
                                          dec_cell, 
                                          dec_embed_inputs, 
                                          target_sequence_length, 
                                          max_target_sequence_length, 
                                          output_layer,
                                          keep_prob)
        
    with tf.variable_scope('inference', reuse=True): 
        
        # Create the inference layer
        infer_pred = decoding_layer_infer(encoder_state,
                                          dec_cell, 
                                          dec_embeddings,
                                          target_vocab_to_int['<GO>'],
                                          target_vocab_to_int['<EOS>'], 
                                          max_target_sequence_length,
                                          target_vocab_size, 
                                          output_layer,
                                          batch_size, 
                                          keep_prob)
        
    return train_pred, infer_pred


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer(decoding_layer)

Tests Passed


### Build the Neural Network
Apply the functions you implemented above to:

- Encode the input using your `encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob,  source_sequence_length, source_vocab_size, encoding_embedding_size)`.
- Process target data using your `process_decoder_input(target_data, target_vocab_to_int, batch_size)` function.
- Decode the encoded input using your `decoding_layer(dec_input, enc_state, target_sequence_length, max_target_sentence_length, rnn_size, num_layers, target_vocab_to_int, target_vocab_size, batch_size, keep_prob, dec_embedding_size)` function.

In [134]:
def seq2seq_model(input_data, target_data, keep_prob, batch_size,
                  source_sequence_length, target_sequence_length,
                  max_target_sentence_length,
                  source_vocab_size, target_vocab_size,
                  enc_embedding_size, dec_embedding_size,
                  rnn_size, num_layers, target_vocab_to_int):
    """
    Build the Sequence-to-Sequence part of the neural network
    :param input_data: Input placeholder
    :param target_data: Target placeholder
    :param keep_prob: Dropout keep probability placeholder
    :param batch_size: Batch Size
    :param source_sequence_length: Sequence Lengths of source sequences in the batch
    :param target_sequence_length: Sequence Lengths of target sequences in the batch
    :param source_vocab_size: Source vocabulary size
    :param target_vocab_size: Target vocabulary size
    :param enc_embedding_size: Decoder embedding size
    :param dec_embedding_size: Encoder embedding size
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: Tuple of (Training BasicDecoderOutput, Inference BasicDecoderOutput)
    """
    
    #rnn_inputs = tf.contrib.layers.embed_sequence(input_data, source_vocab_size, enc_embedding_size)
    
    # Encode the input using the encoding layer
    enc_outputs, enc_state = encoding_layer(input_data, 
                                            rnn_size, 
                                            num_layers, 
                                            keep_prob, 
                                            source_sequence_length, 
                                            source_vocab_size, 
                                            enc_embedding_size)
    
    # Process the data using the process decoder
    dec_input = process_decoder_input(target_data, target_vocab_to_int, batch_size)
    
    #dec_embeddings = tf.Variable(tf.random_uniform([target_vocab_size, dec_embedding_size]))
    #dec_embed_input = tf.nn.embedding_lookup(dec_embeddings, dec_input)
   
    # Decode the embedded input
    Training, Inference = decoding_layer(dec_input, 
                                         enc_state, 
                                         target_sequence_length,
                                         max_target_sequence_length, 
                                         rnn_size, 
                                         num_layers, 
                                         target_vocab_to_int, 
                                         target_vocab_size,
                                         batch_size,
                                         keep_prob,
                                         dec_embedding_size)
    
    return (Training, Inference)

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_seq2seq_model(seq2seq_model)

ValueError: Tensor("LessEqual_6:0", shape=(), dtype=bool) must be from the same graph as Tensor("training/decoder/TrainingHelperInitialize/Equal:0", shape=(?,), dtype=bool).

## Neural Network Training
### Hyperparameters
Tune the following parameters:

- Set `epochs` to the number of epochs.
- Set `batch_size` to the batch size.
- Set `rnn_size` to the size of the RNNs.
- Set `num_layers` to the number of layers.
- Set `encoding_embedding_size` to the size of the embedding for the encoder.
- Set `decoding_embedding_size` to the size of the embedding for the decoder.
- Set `learning_rate` to the learning rate.
- Set `keep_probability` to the Dropout keep probability
- Set `display_step` to state how many steps between each debug output statement

In [144]:
# Number of Epochs
epochs = 10

# Batch Size
batch_size = 512

# RNN Size
rnn_size = 512

# Number of Layers
num_layers = 3

# Embedding Size
encoding_embedding_size = 256
decoding_embedding_size = 256

# Learning Rate
learning_rate = 0.005

# Dropout Keep Probability
keep_probability = 0.5

display_step = 4

### Build the Graph
Build the graph using the neural network you implemented.

In [145]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
save_path = 'checkpoints/dev'
(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()
max_target_sentence_length = max([len(sentence) for sentence in source_int_text])

train_graph = tf.Graph()
with train_graph.as_default():
    input_data, targets, lr, keep_prob, target_sequence_length, max_target_sequence_length, source_sequence_length = model_inputs()

    #sequence_length = tf.placeholder_with_default(max_target_sentence_length, None, name='sequence_length')
    input_shape = tf.shape(input_data)

    train_logits, inference_logits = seq2seq_model(tf.reverse(input_data, [-1]),
                                                   targets,
                                                   keep_prob,
                                                   batch_size,
                                                   source_sequence_length,
                                                   target_sequence_length,
                                                   max_target_sequence_length,
                                                   len(source_vocab_to_int),
                                                   len(target_vocab_to_int),
                                                   encoding_embedding_size,
                                                   decoding_embedding_size,
                                                   rnn_size,
                                                   num_layers,
                                                   target_vocab_to_int)


    training_logits = tf.identity(train_logits.rnn_output, name='logits')
    inference_logits = tf.identity(inference_logits.sample_id, name='predictions')

    masks = tf.sequence_mask(target_sequence_length, max_target_sequence_length, dtype=tf.float32, name='masks')

    with tf.name_scope("optimization"):
        # Loss function
        cost = tf.contrib.seq2seq.sequence_loss(
            training_logits,
            targets,
            masks)

        # Optimizer
        optimizer = tf.train.AdamOptimizer(lr)

        # Gradient Clipping
        gradients = optimizer.compute_gradients(cost)
        capped_gradients = [(tf.clip_by_value(grad, -1., 1.), var) for grad, var in gradients if grad is not None]
        train_op = optimizer.apply_gradients(capped_gradients)


Batch and pad the source and target sequences

In [146]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def pad_sentence_batch(sentence_batch, pad_int):
    """Pad sentences with <PAD> so that each sentence of a batch has the same length"""
    max_sentence = max([len(sentence) for sentence in sentence_batch])
    return [sentence + [pad_int] * (max_sentence - len(sentence)) for sentence in sentence_batch]


def get_batches(sources, targets, batch_size, source_pad_int, target_pad_int):
    """Batch targets, sources, and the lengths of their sentences together"""
    for batch_i in range(0, len(sources)//batch_size):
        start_i = batch_i * batch_size

        # Slice the right amount for the batch
        sources_batch = sources[start_i:start_i + batch_size]
        targets_batch = targets[start_i:start_i + batch_size]

        # Pad
        pad_sources_batch = np.array(pad_sentence_batch(sources_batch, source_pad_int))
        pad_targets_batch = np.array(pad_sentence_batch(targets_batch, target_pad_int))

        # Need the lengths for the _lengths parameters
        pad_targets_lengths = []
        for target in pad_targets_batch:
            pad_targets_lengths.append(len(target))

        pad_source_lengths = []
        for source in pad_sources_batch:
            pad_source_lengths.append(len(source))

        yield pad_sources_batch, pad_targets_batch, pad_source_lengths, pad_targets_lengths


### Train
Train the neural network on the preprocessed data. If you have a hard time getting a good loss, check the forms to see if anyone is having the same problem.

In [147]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
def get_accuracy(target, logits):
    """
    Calculate accuracy
    """
    max_seq = max(target.shape[1], logits.shape[1])
    if max_seq - target.shape[1]:
        target = np.pad(
            target,
            [(0,0),(0,max_seq - target.shape[1])],
            'constant')
    if max_seq - logits.shape[1]:
        logits = np.pad(
            logits,
            [(0,0),(0,max_seq - logits.shape[1])],
            'constant')

    return np.mean(np.equal(target, logits))

# Split data to training and validation sets
train_source = source_int_text[batch_size:]
train_target = target_int_text[batch_size:]
valid_source = source_int_text[:batch_size]
valid_target = target_int_text[:batch_size]
(valid_sources_batch, valid_targets_batch, valid_sources_lengths, valid_targets_lengths ) = next(get_batches(valid_source,
                                                                                                             valid_target,
                                                                                                             batch_size,
                                                                                                             source_vocab_to_int['<PAD>'],
                                                                                                             target_vocab_to_int['<PAD>']))                                                                                                  
with tf.Session(graph=train_graph) as sess:
    sess.run(tf.global_variables_initializer())

    for epoch_i in range(epochs):
        for batch_i, (source_batch, target_batch, sources_lengths, targets_lengths) in enumerate(
                get_batches(train_source, train_target, batch_size,
                            source_vocab_to_int['<PAD>'],
                            target_vocab_to_int['<PAD>'])):

            _, loss = sess.run(
                [train_op, cost],
                {input_data: source_batch,
                 targets: target_batch,
                 lr: learning_rate,
                 target_sequence_length: targets_lengths,
                 source_sequence_length: sources_lengths,
                 keep_prob: keep_probability})


            if batch_i % display_step == 0 and batch_i > 0:


                batch_train_logits = sess.run(
                    inference_logits,
                    {input_data: source_batch,
                     source_sequence_length: sources_lengths,
                     target_sequence_length: targets_lengths,
                     keep_prob: 1.0})


                batch_valid_logits = sess.run(
                    inference_logits,
                    {input_data: valid_sources_batch,
                     source_sequence_length: valid_sources_lengths,
                     target_sequence_length: valid_targets_lengths,
                     keep_prob: 1.0})

                train_acc = get_accuracy(target_batch, batch_train_logits)

                valid_acc = get_accuracy(valid_targets_batch, batch_valid_logits)

                print('Epoch {:>3} Batch {:>4}/{} - Train Accuracy: {:>6.4f}, Validation Accuracy: {:>6.4f}, Loss: {:>6.4f}'
                      .format(epoch_i, batch_i, len(source_int_text) // batch_size, train_acc, valid_acc, loss))

    # Save Model
    saver = tf.train.Saver()
    saver.save(sess, save_path)
    print('Model Trained and Saved')

Epoch   0 Batch    4/269 - Train Accuracy: 0.2916, Validation Accuracy: 0.3604, Loss: 4.5573
Epoch   0 Batch    8/269 - Train Accuracy: 0.2812, Validation Accuracy: 0.3444, Loss: 3.5149
Epoch   0 Batch   12/269 - Train Accuracy: 0.2915, Validation Accuracy: 0.3524, Loss: 3.2904
Epoch   0 Batch   16/269 - Train Accuracy: 0.3302, Validation Accuracy: 0.3618, Loss: 3.0160
Epoch   0 Batch   20/269 - Train Accuracy: 0.3187, Validation Accuracy: 0.3809, Loss: 3.0043
Epoch   0 Batch   24/269 - Train Accuracy: 0.3290, Validation Accuracy: 0.3926, Loss: 2.9658
Epoch   0 Batch   28/269 - Train Accuracy: 0.3250, Validation Accuracy: 0.3964, Loss: 2.9406
Epoch   0 Batch   32/269 - Train Accuracy: 0.3733, Validation Accuracy: 0.4085, Loss: 3.0862
Epoch   0 Batch   36/269 - Train Accuracy: 0.3883, Validation Accuracy: 0.4122, Loss: 2.6605
Epoch   0 Batch   40/269 - Train Accuracy: 0.3865, Validation Accuracy: 0.4356, Loss: 2.6310
Epoch   0 Batch   44/269 - Train Accuracy: 0.4279, Validation Accuracy

Epoch   1 Batch   96/269 - Train Accuracy: 0.6570, Validation Accuracy: 0.6672, Loss: 0.5115
Epoch   1 Batch  100/269 - Train Accuracy: 0.4342, Validation Accuracy: 0.4456, Loss: 0.7168
Epoch   1 Batch  104/269 - Train Accuracy: 0.4993, Validation Accuracy: 0.5312, Loss: 1.4523
Epoch   1 Batch  108/269 - Train Accuracy: 0.5739, Validation Accuracy: 0.5746, Loss: 1.0358
Epoch   1 Batch  112/269 - Train Accuracy: 0.5403, Validation Accuracy: 0.5479, Loss: 0.8436
Epoch   1 Batch  116/269 - Train Accuracy: 0.5959, Validation Accuracy: 0.6060, Loss: 0.7436
Epoch   1 Batch  120/269 - Train Accuracy: 0.6112, Validation Accuracy: 0.6299, Loss: 0.7066
Epoch   1 Batch  124/269 - Train Accuracy: 0.6133, Validation Accuracy: 0.6242, Loss: 0.6409
Epoch   1 Batch  128/269 - Train Accuracy: 0.6478, Validation Accuracy: 0.6307, Loss: 0.6135
Epoch   1 Batch  132/269 - Train Accuracy: 0.6453, Validation Accuracy: 0.6545, Loss: 0.6064
Epoch   1 Batch  136/269 - Train Accuracy: 0.6376, Validation Accuracy

Epoch   2 Batch  188/269 - Train Accuracy: 0.8322, Validation Accuracy: 0.8126, Loss: 0.2074
Epoch   2 Batch  192/269 - Train Accuracy: 0.8448, Validation Accuracy: 0.8249, Loss: 0.2106
Epoch   2 Batch  196/269 - Train Accuracy: 0.8394, Validation Accuracy: 0.8197, Loss: 0.1996
Epoch   2 Batch  200/269 - Train Accuracy: 0.8335, Validation Accuracy: 0.8311, Loss: 0.2096
Epoch   2 Batch  204/269 - Train Accuracy: 0.8443, Validation Accuracy: 0.8511, Loss: 0.1982
Epoch   2 Batch  208/269 - Train Accuracy: 0.8430, Validation Accuracy: 0.8469, Loss: 0.2005
Epoch   2 Batch  212/269 - Train Accuracy: 0.8326, Validation Accuracy: 0.8554, Loss: 0.1939
Epoch   2 Batch  216/269 - Train Accuracy: 0.8375, Validation Accuracy: 0.8404, Loss: 0.2035
Epoch   2 Batch  220/269 - Train Accuracy: 0.8460, Validation Accuracy: 0.8421, Loss: 0.1758
Epoch   2 Batch  224/269 - Train Accuracy: 0.8447, Validation Accuracy: 0.8453, Loss: 0.1969
Epoch   2 Batch  228/269 - Train Accuracy: 0.8635, Validation Accuracy

Epoch   4 Batch   16/269 - Train Accuracy: 0.9331, Validation Accuracy: 0.9283, Loss: 0.0623
Epoch   4 Batch   20/269 - Train Accuracy: 0.9347, Validation Accuracy: 0.9289, Loss: 0.0535
Epoch   4 Batch   24/269 - Train Accuracy: 0.9269, Validation Accuracy: 0.9287, Loss: 0.0546
Epoch   4 Batch   28/269 - Train Accuracy: 0.9170, Validation Accuracy: 0.9410, Loss: 0.0571
Epoch   4 Batch   32/269 - Train Accuracy: 0.9374, Validation Accuracy: 0.9276, Loss: 0.0470
Epoch   4 Batch   36/269 - Train Accuracy: 0.9314, Validation Accuracy: 0.9239, Loss: 0.0532
Epoch   4 Batch   40/269 - Train Accuracy: 0.9262, Validation Accuracy: 0.9365, Loss: 0.0541
Epoch   4 Batch   44/269 - Train Accuracy: 0.9390, Validation Accuracy: 0.9253, Loss: 0.0570
Epoch   4 Batch   48/269 - Train Accuracy: 0.9461, Validation Accuracy: 0.9285, Loss: 0.0482
Epoch   4 Batch   52/269 - Train Accuracy: 0.9349, Validation Accuracy: 0.9335, Loss: 0.0452
Epoch   4 Batch   56/269 - Train Accuracy: 0.9377, Validation Accuracy

Epoch   5 Batch  108/269 - Train Accuracy: 0.9621, Validation Accuracy: 0.9577, Loss: 0.0310
Epoch   5 Batch  112/269 - Train Accuracy: 0.9559, Validation Accuracy: 0.9503, Loss: 0.0346
Epoch   5 Batch  116/269 - Train Accuracy: 0.9587, Validation Accuracy: 0.9463, Loss: 0.0296
Epoch   5 Batch  120/269 - Train Accuracy: 0.9517, Validation Accuracy: 0.9602, Loss: 0.0341
Epoch   5 Batch  124/269 - Train Accuracy: 0.9582, Validation Accuracy: 0.9559, Loss: 0.0287
Epoch   5 Batch  128/269 - Train Accuracy: 0.9602, Validation Accuracy: 0.9395, Loss: 0.0297
Epoch   5 Batch  132/269 - Train Accuracy: 0.9492, Validation Accuracy: 0.9635, Loss: 0.0321
Epoch   5 Batch  136/269 - Train Accuracy: 0.9464, Validation Accuracy: 0.9474, Loss: 0.0342
Epoch   5 Batch  140/269 - Train Accuracy: 0.9531, Validation Accuracy: 0.9566, Loss: 0.0344
Epoch   5 Batch  144/269 - Train Accuracy: 0.9635, Validation Accuracy: 0.9579, Loss: 0.0248
Epoch   5 Batch  148/269 - Train Accuracy: 0.9704, Validation Accuracy

Epoch   6 Batch  200/269 - Train Accuracy: 0.9703, Validation Accuracy: 0.9566, Loss: 0.0208
Epoch   6 Batch  204/269 - Train Accuracy: 0.9599, Validation Accuracy: 0.9556, Loss: 0.0223
Epoch   6 Batch  208/269 - Train Accuracy: 0.9704, Validation Accuracy: 0.9693, Loss: 0.0243
Epoch   6 Batch  212/269 - Train Accuracy: 0.9620, Validation Accuracy: 0.9667, Loss: 0.0269
Epoch   6 Batch  216/269 - Train Accuracy: 0.9596, Validation Accuracy: 0.9587, Loss: 0.0324
Epoch   6 Batch  220/269 - Train Accuracy: 0.9625, Validation Accuracy: 0.9596, Loss: 0.0211
Epoch   6 Batch  224/269 - Train Accuracy: 0.9627, Validation Accuracy: 0.9600, Loss: 0.0277
Epoch   6 Batch  228/269 - Train Accuracy: 0.9676, Validation Accuracy: 0.9641, Loss: 0.0202
Epoch   6 Batch  232/269 - Train Accuracy: 0.9489, Validation Accuracy: 0.9621, Loss: 0.0248
Epoch   6 Batch  236/269 - Train Accuracy: 0.9712, Validation Accuracy: 0.9672, Loss: 0.0192
Epoch   6 Batch  240/269 - Train Accuracy: 0.9773, Validation Accuracy

Epoch   8 Batch   28/269 - Train Accuracy: 0.9736, Validation Accuracy: 0.9687, Loss: 0.0166
Epoch   8 Batch   32/269 - Train Accuracy: 0.9807, Validation Accuracy: 0.9709, Loss: 0.0136
Epoch   8 Batch   36/269 - Train Accuracy: 0.9695, Validation Accuracy: 0.9678, Loss: 0.0159
Epoch   8 Batch   40/269 - Train Accuracy: 0.9771, Validation Accuracy: 0.9702, Loss: 0.0182
Epoch   8 Batch   44/269 - Train Accuracy: 0.9712, Validation Accuracy: 0.9664, Loss: 0.0186
Epoch   8 Batch   48/269 - Train Accuracy: 0.9727, Validation Accuracy: 0.9738, Loss: 0.0149
Epoch   8 Batch   52/269 - Train Accuracy: 0.9817, Validation Accuracy: 0.9705, Loss: 0.0108
Epoch   8 Batch   56/269 - Train Accuracy: 0.9731, Validation Accuracy: 0.9708, Loss: 0.0146
Epoch   8 Batch   60/269 - Train Accuracy: 0.9760, Validation Accuracy: 0.9727, Loss: 0.0184
Epoch   8 Batch   64/269 - Train Accuracy: 0.9845, Validation Accuracy: 0.9751, Loss: 0.0122
Epoch   8 Batch   68/269 - Train Accuracy: 0.9651, Validation Accuracy

Epoch   9 Batch  120/269 - Train Accuracy: 0.9732, Validation Accuracy: 0.9709, Loss: 0.0161
Epoch   9 Batch  124/269 - Train Accuracy: 0.9869, Validation Accuracy: 0.9696, Loss: 0.0118
Epoch   9 Batch  128/269 - Train Accuracy: 0.9775, Validation Accuracy: 0.9622, Loss: 0.0142
Epoch   9 Batch  132/269 - Train Accuracy: 0.9665, Validation Accuracy: 0.9692, Loss: 0.0146
Epoch   9 Batch  136/269 - Train Accuracy: 0.9760, Validation Accuracy: 0.9738, Loss: 0.0156
Epoch   9 Batch  140/269 - Train Accuracy: 0.9830, Validation Accuracy: 0.9718, Loss: 0.0154
Epoch   9 Batch  144/269 - Train Accuracy: 0.9879, Validation Accuracy: 0.9726, Loss: 0.0115
Epoch   9 Batch  148/269 - Train Accuracy: 0.9804, Validation Accuracy: 0.9679, Loss: 0.0127
Epoch   9 Batch  152/269 - Train Accuracy: 0.9817, Validation Accuracy: 0.9671, Loss: 0.0127
Epoch   9 Batch  156/269 - Train Accuracy: 0.9858, Validation Accuracy: 0.9743, Loss: 0.0129
Epoch   9 Batch  160/269 - Train Accuracy: 0.9814, Validation Accuracy

### Save Parameters
Save the `batch_size` and `save_path` parameters for inference.

In [152]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# Save parameters for checkpoint
helper.save_params(save_path)

# Checkpoint

In [153]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import tensorflow as tf
import numpy as np
import helper
import problem_unittests as tests

_, (source_vocab_to_int, target_vocab_to_int), (source_int_to_vocab, target_int_to_vocab) = helper.load_preprocess()
load_path = helper.load_params()

## Sentence to Sequence
To feed a sentence into the model for translation, you first need to preprocess it.  Implement the function `sentence_to_seq()` to preprocess new sentences.

- Convert the sentence to lowercase
- Convert words into ids using `vocab_to_int`
 - Convert words not in the vocabulary, to the `<UNK>` word id.

In [154]:
def sentence_to_seq(sentence, vocab_to_int):
    """
    Convert a sentence to a sequence of ids
    :param sentence: String
    :param vocab_to_int: Dictionary to go from the words to an id
    :return: List of word ids
    """
    
    # Convert all chars in the sentence to lower-case
    sentence_lc   = sentence.lower()
    
    # Replace unknown words with <UNK>
    unknown_words = vocab_to_int["<UNK>"]
    
    # Convert the individual words of the sentence to their corresponding ids
    word_ids = [vocab_to_int.get(word, unknown_words) for word in sentence_lc.split()]
    
    return word_ids


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_sentence_to_seq(sentence_to_seq)

Tests Passed


## Translate
This will translate `translate_sentence` from English to French.

In [155]:
translate_sentence = 'he saw a old yellow truck .'


"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
translate_sentence = sentence_to_seq(translate_sentence, source_vocab_to_int)

loaded_graph = tf.Graph()
with tf.Session(graph=loaded_graph) as sess:
    # Load saved model
    loader = tf.train.import_meta_graph(load_path + '.meta')
    loader.restore(sess, load_path)

    input_data = loaded_graph.get_tensor_by_name('input:0')
    logits = loaded_graph.get_tensor_by_name('predictions:0')
    target_sequence_length = loaded_graph.get_tensor_by_name('target_sequence_length:0')
    source_sequence_length = loaded_graph.get_tensor_by_name('source_sequence_length:0')
    keep_prob = loaded_graph.get_tensor_by_name('keep_prob:0')

    translate_logits = sess.run(logits, {input_data: [translate_sentence]*batch_size,
                                         target_sequence_length: [len(translate_sentence)*2]*batch_size,
                                         source_sequence_length: [len(translate_sentence)]*batch_size,
                                         keep_prob: 1.0})[0]

print('Input')
print('  Word Ids:      {}'.format([i for i in translate_sentence]))
print('  English Words: {}'.format([source_int_to_vocab[i] for i in translate_sentence]))

print('\nPrediction')
print('  Word Ids:      {}'.format([i for i in translate_logits]))
print('  French Words: {}'.format(" ".join([target_int_to_vocab[i] for i in translate_logits])))


INFO:tensorflow:Restoring parameters from checkpoints/dev
Input
  Word Ids:      [72, 53, 24, 60, 83, 101, 131]
  English Words: ['he', 'saw', 'a', 'old', 'yellow', 'truck', '.']

Prediction
  Word Ids:      [351, 7, 299, 77, 43, 128, 311, 1]
  French Words: il a conduit un camion jaune . <EOS>


## Imperfect Translation
You might notice that some sentences translate better than others.  Since the dataset you're using only has a vocabulary of 227 English words of the thousands that you use, you're only going to see good results using these words.  For this project, you don't need a perfect translation. However, if you want to create a better translation model, you'll need better data.

You can train on the [WMT10 French-English corpus](http://www.statmt.org/wmt10/training-giga-fren.tar).  This dataset has more vocabulary and richer in topics discussed.  However, this will take you days to train, so make sure you've a GPU and the neural network is performing well on dataset we provided.  Just make sure you play with the WMT10 corpus after you've submitted this project.
## Submitting This Project
When submitting this project, make sure to run all the cells before saving the notebook. Save the notebook file as "dlnd_language_translation.ipynb" and save it as a HTML file under "File" -> "Download as". Include the "helper.py" and "problem_unittests.py" files in your submission.